src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp

```diff
@@ -0,0 +1,443 @@
+        break;
+    case RemoveClientCertificate::Id:
+        CommandHandlerInterface::HandleCommand<RemoveClientCertificate::DecodableType>(
+            ctx, [this](HandlerContext & innerCtx, const auto & req) { HandleRemoveClientCertificate(innerCtx, req); });
+        break;
+    }
+}
+
+void TlsCertificateManagementServer::HandleProvisionRootCertificate(HandlerContext & ctx,
+                                                                    const ProvisionRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: ProvisionRootCertificate");
+
+    VerifyOrReturn(req.certificate.size() <= kSpecMaxCertBytes,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    VerifyOrReturn(req.caid.IsNull() || req.caid.Value() <= kMaxRootCertId,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+
+    DataModel::Nullable<Tls::TLSCAID> foundId;
+    auto lookupResult = mDelegate.LookupRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(),
+                                                 req.certificate, [&](auto & certificate) -> CHIP_ERROR {
+                                                     foundId = certificate.caid;
+                                                     return CHIP_NO_ERROR;
+                                                 });
+    if (lookupResult != CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::AlreadyExists);
+        return;
+    }
+
+    ProvisionRootCertificateResponse::Type response;
+    auto status = mDelegate.ProvisionRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req,
+                                              response.caid);
+
+    if (!status.IsSuccess())
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    VerifyOrDieWithMsg(response.caid <= kMaxRootCertId, NotSpecified, "Spec requires CAID to be < kMaxRootCertId");
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+}
+
+void TlsCertificateManagementServer::HandleFindRootCertificate(HandlerContext & ctx, const FindRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: FindRootCertificate");
+    CHIP_ERROR result;
+    if (req.caid.IsNull())
+    {
+        result = mDelegate.RootCertsForFabric(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(),
+                                              [&](auto & certs) -> CHIP_ERROR {
+                                                  FindRootCertificateResponse::Type response;
+                                                  response.certificateDetails = certs;
+                                                  ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
```

Comment:
This does not seem to match spec (which may have been updated to fix a security/privacy bug since you wrote this code)....